source/common/router/header_formatter.cc

```diff
@@ -114,6 +123,43 @@ parseUpstreamMetadataField(absl::string_view params_str) {
   };
 }
 
+// Parses the parameters for PER_REQUEST_STATE and returns a function suitable for accessing the
+// specified metadata from an RequestInfo::RequestInfo. Expects a string formatted as:
+//   (<state_name>)
+// The state name is expected to be in reverse DNS format, though this is not enforced by
+// this function.
+std::function<std::string(const Envoy::RequestInfo::RequestInfo&)>
+parseFilterStateField(absl::string_view param_str) {
+  absl::string_view modified_param_str = StringUtil::trim(param_str);
+  if (modified_param_str.empty() || modified_param_str.front() != '(' ||
+      modified_param_str.back() != ')') {
+    throw EnvoyException(formatFilterStateParseException(param_str));
+  }
+  modified_param_str = modified_param_str.substr(1, modified_param_str.size() - 2); // trim parens
+  if (modified_param_str.size() == 0) {
+    throw EnvoyException(formatFilterStateParseException(param_str));
+  }
+
+  std::string param(modified_param_str);
+  return [param](const Envoy::RequestInfo::RequestInfo& request_info) -> std::string {
+    const Envoy::RequestInfo::FilterState& dynamic_metadata = request_info.perRequestState();
+
+    // No such value means don't output anything.
+    if (!dynamic_metadata.hasDataWithName(param)) {
+      return std::string();
+    }
+
+    // Value exists but isn't string accessible is a contract violation; throw an error.
+    if (!dynamic_metadata.hasData<StringAccessor>(param)) {
+      throw EnvoyException(fmt::format("Invalid header information: PER_REQUEST_STATE value \"{}\" "
+                                       "exists but is not string accessible",
+                                       param));
```

Comment:
Throwing an exception here seems pretty aggressive. I'm not even sure it'll get converted into a 5xx since the exceptions that the ConnectionManagerImpl catches are subclasses of EnvoyException. Is it not sufficient write to the debug log and return an empty string?

Reply:
Sure.  One of the things I was looking for was guidance on error handling; I should have said.  I will convert.